loadtest/app.go

```diff
@@ -0,0 +1,183 @@
+	app.config.SetDefault("loadtest.requests.period.ms", 0)
+	app.setPlayerConfigurationDefaults()
+	app.setClanConfigurationDefaults()
+	app.setMembershipConfigurationDefaults()
+}
+
+func (app *App) loadConfiguration(configFile string) {
+	l := app.logger.With(
+		zap.String("source", "loadtest/app"),
+		zap.String("operation", "loadConfiguration"),
+		zap.String("configFile", configFile),
+	)
+
+	app.config.SetConfigType("yaml")
+	app.config.SetConfigFile(configFile)
+	app.config.SetEnvPrefix("khan")
+	app.config.AddConfigPath(".")
+	app.config.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	app.config.AutomaticEnv()
+
+	if err := app.config.ReadInConfig(); err == nil {
+		log.I(l, "Loaded config file successfully.")
+	} else {
+		log.P(l, "Config file failed to load.")
+	}
+}
+
+func (app *App) configureOperations() {
+	app.configurePlayerOperations()
+	app.configureClanOperations()
+	app.configureMembershipOperations()
+}
+
+func (app *App) configureCache(sharedClansFile string) {
+	var err error
+	app.cache, err = getCacheImpl(app.config, sharedClansFile)
+	if err != nil {
+		l := app.logger.With(
+			zap.String("source", "loadtest/app"),
+			zap.String("operation", "configureCache"),
+			zap.String("sharedClansFile", sharedClansFile),
+		)
+		log.P(l, "Error reading shared clans config.", func(cm log.CM) {
+			cm.Write(zap.String("error", err.Error()))
+		})
+	}
+}
+
+func (app *App) configureClient() {
+	app.client = lib.NewKhan(app.config)
+}
+
+// Run executes the load test suite
+func (app *App) Run() error {
+	nRequests := app.config.GetInt("loadtest.requests.amount")
+	periodMs := app.config.GetInt("loadtest.requests.period.ms")
+	for i := 0; i < nRequests; i++ {
+		err := app.performOperation()
+		if err != nil {
+			return err
+		}
+		if i+1 < nRequests {
+			time.Sleep(time.Duration(periodMs) * time.Millisecond)
+		}
+	}
+	return nil
+}
+
+func (app *App) performOperation() error {
+	operation, err := app.getRandomOperation()
+	if err != nil {
+		return err
+	}
+	err = operation.execute()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (app *App) getRandomOperation() (operation, error) {
+	sampleSpace, err := app.getOperationSampleSpace()
+	if err != nil {
+		return operation{}, err
+	}
+	return getRandomOperationFromSampleSpace(sampleSpace, -1)
+}
+
+func (app *App) getOperationSampleSpace() ([]operation, error) {
+	var pSum float64
+	var sampleSpace []operation
+	for _, operation := range app.operations {
+		ok, err := operation.canExecute()
+		if err != nil {
+			return nil, err
+		}
+		if ok && operation.probability > 0 {
+			pSum += operation.probability
+			sampleSpace = append(sampleSpace, operation)
+		}
+	}
+	if pSum <= 0 {
+		return nil, &GenericError{"NoOperationsAvailableError", "No operations can be executed anymore."}
+	}
+	normalizeSampleSpace(sampleSpace, pSum)
+	return sampleSpace, nil
+}
+
+func normalizeSampleSpace(sampleSpace []operation, pSum float64) {
```

Comment:
It seems weird to me that you need to pass pSum as an argument here as it can be calculated from sampleSpace, right? I know you'd need to iterate over the array twice, but IMO it makes more sense to have this encapsulated here

Reply:
Makes sense. What about leaving it as an optional parameter? If negative, the function should compute it.

Reply:
What are you gaining from sending it as an optional parameter?

Reply:
Not looping again over the operations. Actually, I'm thinking there's no need for a function like this `normalizeSampleSpace()` function, I'm seriously thinking about moving this three-liner to the only function that calls it.

Reply:
The function is not all that generic to actually be a separated function, there's no structure representing a generic sample space of probabilities (to represent not only Khan's load test operations sample space, but every possible sample space someone could think about). I guess I'll go without it! What you think?

Reply:
I don't see the need for a specific function either. Let's go without it then.